File: java-tests/src/test/java/tests/environment/main/case2/service/ProductServiceTest.java
package tests.environment.main.case2.service;

import static org.assertj.core.api.AssertionsForInterfaceTypes.assertThat;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.boot.test.system.CapturedOutput;
import org.springframework.boot.test.system.OutputCaptureExtension;
import tests.case2.service.ProductService;

@ExtendWith(OutputCaptureExtension.class)
@SpringBootTest(
        properties = {
                "spring.jpa.properties.hibernate.detector.enabled=true",
                "spring.jpa.properties.hibernate.detector.threshold=2",
                "spring.jpa.properties.hibernate.detector.exclude=",
                "spring.jpa.properties.hibernate.detector.level=warn"
        })
public class ProductServiceTest {

    @Autowired
    private ProductService productService;

    @Test
    @DisplayName("다(N)를 조회하는 상황에서 감지한다.")
    void testFindAllProducts(CapturedOutput output) {
        productService.findAllProducts();

        assertThat(output).contains("N+1 query detected");
    }
}

File: kotlin-tests/src/test/kotlin/tests/environment/main/service/ProductServiceTest.kt
package tests.environment.main.service

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.system.CapturedOutput
import org.springframework.boot.test.system.OutputCaptureExtension
import tests.service.ProductService

@ExtendWith(OutputCaptureExtension::class)
@SpringBootTest(
    properties = [
        "nplus1detector.enabled=true",
        "nplus1detector.threshold=2",
        "nplus1detector.exclude=",
        "nplus1detector.level=warn"
    ]
)
class ProductServiceTest {

    @Autowired
    private lateinit var productService: ProductService

    @Test
    @DisplayName("다(N)를 조회하는 상황에서 감지한다.")
    fun testFindAllProducts(output: CapturedOutput) {
        productService.findAllProducts()

        assertThat(output).contains("N+1 query detected")
    }
}
